sdk/core/azure-core/azure/core/pipeline/transport/_requests_basic.py

```diff
@@ -119,39 +154,58 @@ def __next__(self):
         retry_active = True
         retry_total = 3
         retry_interval = 1  # 1 second
-        while retry_active:
-            try:
-                chunk = next(self.iter_content_func)
-                if not chunk:
-                    raise StopIteration()
-                self.downloaded += self.block_size
-                return chunk
-            except StopIteration:
-                self.response.internal_response.close()
+        try:
+            chunk = next(self.iter_content_func)
+            if not chunk:
                 raise StopIteration()
-            except (requests.exceptions.ChunkedEncodingError,
-                    requests.exceptions.ConnectionError):
+            self.downloaded += self.block_size
+            return chunk
+        except StopIteration:
+            self.response.internal_response.close()
+            raise StopIteration()
+        except (requests.exceptions.ChunkedEncodingError,
+                requests.exceptions.ConnectionError) as ex:
+            while retry_active:
                 retry_total -= 1
                 if retry_total <= 0:
-                    retry_active = False
+                    _LOGGER.warning("Unable to stream download: %s", ex)
+                    raise ex
                 else:
                     time.sleep(retry_interval)
-                    headers = {'range': 'bytes=' + str(self.downloaded) + '-'}
-                    resp = self.pipeline.run(self.request, stream=True, headers=headers)
-                    if resp.http_response.status_code == 416:
-                        raise
-                    chunk = next(self.iter_content_func)
+                    # todo handle pre-set range & x-ms-range
+                    headers = self.request.headers
+                    if not self.range_header:
+                        range_header = {'range': 'bytes=' + str(self.downloaded) + '-'}
+                    else:
+                        range_header = {self.range_header: make_range_header(self.range, self.downloaded)}
+                    if self.etag:
```

Comment:
Combining the responses from multiple requests with different ranges must be done only if there is a strong validator, such as a non-weak etag (cf. section 3.3 of RFC 7234 ("Combining Partial Content") and section 4.3. of RFC 7233 ("Combining Ranges")).
So to be on the safe side, if there is no (non-weak) etag, there should be no retry.

(To see why, imagine a server response that dynamically generates some json data that does not guarantee the same order of keys in the json objects on the next request: this is a valid server behavior, but the client here would produce corrupt data).